### ISSUE TITLE
CC: Windows IDTech failing on SigCapture

### ISSUE THREAD

**Issue by joelbrock** [CORE-POS/IS4C#436]
Driver compiled with auto_state_change = false 
Paycard config set to either direct OR messages

Symptom is the same as we've seen before, authorization occurs, requests customer signature, PINpad is signed, PINpad proceeds to "approved", POS screen shows:
![screen shot 2015-01-21 at 10 27 29 am](https://cloud.githubusercontent.com/assets/52835/5842255/21f2db30-a158-11e4-897a-47c3e90de59d.png)

pos.exe -v output is as follows:
```
Received: termSig
Full Report 9
2 3 0 7a 46 1f 23 df 3

DMSG: -1: 6
The driver has become confused!
Full Report 9
2 3 0 7a 46 19 25 d9 3

DMSG: -1: 6
The driver has become confused!
Full Report 12
2 6 0 8a 46 20 ff ff ff 13 ed 3

DMSG: -1: 6
The driver has become confused!
Full Report 17
2 b 0 8a 46 22 0 0 0 0 3f 1 ef 0 3f 21
3

DMSG: -1: 6
The driver has become confused!
Full Report 15
2 9 0 8a 46 40 e c 0 0 0 3 8d 2d 3

DMSG: -1: 6
The driver has become confused!
Full Report 12
2 6 0 8a 46 41 0 0 0 8d 11 3

DMSG: -1: 6
The driver has become confused!
Full Report 15
2 9 0 8a 46 4f 1 0 1 0 0 0 83 21 3

DMSG: -1: 15 e1 0
The driver has become confused!
Full Report 44
2 26 0 7a 46 4 2 11 f 5 0 1c 0 73 0 49
0 16 0 10 12 1 0 0 4 0 0 0 1 ee 0 0
f 0 7 0 43 6c 65 61 72 a8 ec 3

DMSG: -1: 6
The driver has become confused!
Full Report 43
2 25 0 7a 46 4 1 11 f cc 0 1c 0 3a 1 49
0 15 0 10 12 1 0 0 4 0 0 0 1 0 bb 0
17 0 7 0 44 6f 6e 65 1d ed 3

DMSG: -1: 6
The driver has become confused!
Full Report 15
2 9 0 8a 46 40 e c 0 0 0 3 8d 2d 3

DMSG: -1: 6
The driver has become confused!
Full Report 26
2 14 0 8a 46 4f 64 0 53 0 b 0 70 6c 65 61
73 65 20 73 69 67 6e 82 2c 3

DMSG: -1: 6
The driver has become confused!
Full Report 21
2 f 0 7a 46 3 5 0 76 0 36 1 e6 0 f 0
0 0 92 6a 3

DMSG: -1: 6
The driver has become confused!
Full Report 17
2 b 0 7a 46 10 5 7a 0 0 0 ff ff ff ac 4c
3

DMSG: -1: 6
The driver has become confused!
Received: termSig
Full Report 9
2 3 0 7a 46 1f 23 df 3

DMSG: -1: 6
The driver has become confused!
Full Report 9
2 3 0 7a 46 19 25 d9 3

DMSG: -1: 6
The driver has become confused!
Full Report 12
2 6 0 8a 46 20 ff ff ff 13 ed 3

DMSG: -1: 6
The driver has become confused!
Full Report 17
2 b 0 8a 46 22 0 0 0 0 3f 1 ef 0 3f 21
3

DMSG: -1: 6
The driver has become confused!
Full Report 15
2 9 0 8a 46 40 e c 0 0 0 3 8d 2d 3

DMSG: -1: 6
The driver has become confused!
Full Report 12
2 6 0 8a 46 41 0 0 0 8d 11 3

DMSG: -1: 6
The driver has become confused!
Full Report 15
2 9 0 8a 46 4f 1 0 1 0 0 0 83 21 3

DMSG: -1: 15 e1 0
The driver has become confused!
Full Report 44
2 26 0 7a 46 4 2 11 f 5 0 1c 0 73 0 49
0 16 0 10 12 1 0 0 4 0 0 0 1 ee 0 0
f 0 7 0 43 6c 65 61 72 a8 ec 3

DMSG: -1: 6
The driver has become confused!
Full Report 43
2 25 0 7a 46 4 1 11 f cc 0 1c 0 3a 1 49
0 15 0 10 12 1 0 0 4 0 0 0 1 0 bb 0
17 0 7 0 44 6f 6e 65 1d ed 3

DMSG: -1: 6
The driver has become confused!
Full Report 15
2 9 0 8a 46 40 e c 0 0 0 3 8d 2d 3

DMSG: -1: 6
The driver has become confused!
Full Report 26
2 14 0 8a 46 4f 64 0 53 0 b 0 70 6c 65 61
73 65 20 73 69 67 6e 82 2c 3

DMSG: -1: 6
The driver has become confused!
Full Report 21
2 f 0 7a 46 3 5 0 76 0 36 1 e6 0 f 0
0 0 92 6a 3

DMSG: -1: 6
The driver has become confused!
Full Report 17
2 b 0 7a 46 10 5 7a 0 0 0 ff ff ff ac 4c
3

DMSG: -1: 6
The driver has become confused!
DMSG: 7: 7a 1 7f 30
Full Report 12
2 6 0 8a 46 20 ff ff ff 13 ed 3

Full Report 17
2 b 0 8a 46 22 0 0 1b 0 3f 1 73 0 b8 c0
3

Full Report 15
2 9 0 8a 46 40 e c 0 0 0 3 8d 2d 3

Full Report 12
2 6 0 8a 46 41 0 0 0 8d 11 3

Full Report 35
2 1d 0 8a 46 4f 28 0 53 0 14 0 61 70 70 72
6f 76 65 64 20 2d 20 74 68 61 6e 6b 20 79 6f 75
f1 f 3

Full Report 9
2 3 0 7a 46 23 1f e3 3

DMSG: 7: ff ff ff ff ff ff ff ff df ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff df ff 8
9 e5 3
```

**Comment by gohanman** [comment]
Hmmm. That looks relatively OK. The "confused" messages aren't helpful in this case. 

As general info, signature capture is *always* initiated by POS regardless of auto_state_change (because the driver doesn't know about settings like "no signature required under $25").

The "DMSG: 7: 7a 1 7f 30" line is the device indicating the user pressed Done. The "2 3 0 7a 46 23 1f e3 3" line is the driver asking the device for the signature image (the stuff in between is drawing some things on the screen).

That very last message "DMSG: 7: ff ff ff ff ff ff ff ff df ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff df ff 8
9 e5 3" looks like the tail end of a bitmap but the beginning of it is missing for some reason.

This fixes it for me: dc8dbb92ce72b1c8f32f874290e4f92e9ff271e0
**NB: I don't yet know what the linux impact of this change is**

All that does is make the pause after writing to the device shorter. My *guess* is there's a limited intermediate buffer at the OS level. If the device is trying to write more data than that buffer can hold, the driver needs to be reading simultaneously to avoid missing anything.

**Comment by joelbrock** [comment]
this appears to have resolved the issue on the Harvest WIndows testlane.  THanks @gohanman
